### ISSUE TITLE
Type narrowing from a generic type

### ISSUE THREAD

**Issue by neutrinoceros** [microsoft/pyright#10153]
**Is your feature request related to a problem? Please describe.**

I'm trying to check a function that takes a generic type `Array[T]` as input and outputs, and dispatches internally to specialized functions depending on the runtime value of `T`. `pyright`, as of version 1.1.397, does not appear to narrow `T` correctly, leading to spurious errors.

A real life example, together with simplified examples not involving numpy, can be found at https://github.com/numpy/numpy/issues/28572

In particular, see [this playground](https://pyright-play.net/?strict=true&code=MYGwhgzhAEAmAuBPADgUwPoDMDMAmAXNAHQkBQokMCKGmAbACyElGnnhTQCCATj2IgDaAFQC6%2BUtCnQAAsh4B7NDySTpsVJjhI0ACgioQmAJTQAtAD5ow5mVIatmXWEK9%2BQ6mix5oAH200WIyixhLS0ACWWhEQEQB2EPBgccCozkSeqAA0AV44uKFq4VI8qABuqGAg6DppYKbQAMRSbgKCmd64frm0wUA)

**Describe the solution you’d like**

I'd like `pyright` to narrow the type of `a` from `Array[dtype_f32 | dtype_f64]` to `#  Array[dtype_f32]` in the linked playground example.

**Comment by erictraut** [comment]
Narrowing in the manner that you've suggested wouldn't be safe in general. Checking the type of a single attribute (or in this case, the return type of a property getter) doesn't mean that the type argument for a generic type can be narrowed.

You could create a user-defined type guard function if you know this will be safe for your type.

```python
from typing import Any, TypeIs


class dtype_f32: ...
class dtype_f64: ...

class Array[T]:
    @property
    def dtype(self) -> T: ...

def narrow_array[T](a: Array[Any], b: type[T]) -> TypeIs[Array[T]]:
    return isinstance(a.dtype, b)

def f(a: Array[dtype_f32 | dtype_f64]):
    if narrow_array(a, dtype_f32):
        reveal_type(a)
```

**Comment by jorenham** [comment]
> Narrowing in the manner that you've suggested wouldn't be safe in general. Checking the type of a single attribute (or in this case, the return type of a property getter) doesn't mean that the type argument for a generic type can be narrowed.

Would it be safe if the property is `@final`?

**Comment by erictraut** [comment]
> Would it be safe if the property is @final?

No, it wouldn't be safe.

Safety depends on how the class happens to interpret the meaning of the type parameter — something that a static type checker would not be able to discern from type information provided to it.

Consider the following implementation of `Array`. This is admittedly contrived, but it demonstrates how such an assumption would fail.

```python
from typing import reveal_type, Sequence

class dtype_f32: ...
class dtype_f64: ...

class Array[T]:
    def __init__(self, dtypes: Sequence[T]):
        self._dtypes = dtypes
        self._index = 0

    @property
    def dtype(self) -> T:
        x = self._dtypes[self._index] 
        self._index = (self._index + 1) % len(self._dtypes)
        return x

a = Array([dtype_f32(), dtype_f64()])
reveal_type(a)

d1 = a.dtype
print(d1) # dtype_f32
d2 = a.dtype
print(d2) # dtype_f64
```

You may happen to know that this narrowing would be safe because you understand the internals of the `Array` class and how it uses the type parameter, but a type checker can't make this assumption based on an `isinstance` check.

**Comment by neutrinoceros** [comment]
Thank you both. Unfortunately, the proposed solution using `TypeIs` doesn't seem to scale to my actual use case; the key difference is on the input type to `f`, which uses a `TypeVar`.

See [this playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQDRQCCKcjAKooQJIDOjZXr0J4qKEiIBQkgMYVBvKABNahKsADMAJgBcUAHSHZ8octXqAbABY9h-dLkLmIEGTgBtdgF0dkqP6gAAQRwOjw4PwClQmAzLgAKYQpgAEooAFoAPih2WyNJaNiAC0EqFS5PL3iyPSYXN3cWOC9Gcro9VUq0rJyuPkb6j28fSP8iGABXEBQoMn02wigAXiW4umlCqGBPPXiF9W1W82BrFKqa51chr27suqvK3wDMYtKF6qOuA60Up%2BeAoikSg0BJkFKjf7jKYzMgQwgUJCvXhlVQfNZqE5WX4Q56A8jUVFgnEAwiTaazOEUYR-f6zIQiGBiCQgaopIA) for illustration.

pyright appears to mark my 2 useful branches as unreachable. Is this also working as intended ? Am I doing anything wrong here ? Should I open a new issue ?

**Comment by erictraut** [comment]
You're using a value-constrained type variable in your example. I recommend avoiding these if you can help it because their behavior is not well defined, and it's not clear to me that we'll ever be able to standardize their behavior in a way that composes well with other features of the type system. An overload is almost always a preferable way to handle polymorphism in a function. See [here](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQjEmVkD68Cx9AginPQBVExAJIBnepTFjSMViiakAUEoDGZKWKjUOxVsADMAJgBcUAHSW1G6dt36AbABYzl8yvWao3ECEpwAbQEAXRMlKAioAAEEcE48OHDI6mJgO2EAChkyYABKKABaAD4oAVcrJRS0gAspVh1hIOCMyjMfP0DeOGD6Bs4zXSb84tLhcQD2-ybQpIjGGABXEBQoSnM%2B4igAXi30zhUoolIKGkrUqGAWtt8pjf1jYOGSycC7wyNQiytDknIqajOaUurW8N1e9mAzkehWeYICbyh5XcVQuV1BHXhEIeUAAPujbhCoU98eDhPcPriSZiyZCnDNIpganUNi1elijLkwgyGYxmGxdC1crNufMlitKMLiGQkEyxPUBZQ2TTnJzhTzFCx2JlKELuZFRctVpKyDIuXrVtJZPJFCBBUogA) for a working example.

> Am I doing anything wrong here ? Should I open a new issue ?
At first glance, this does look like a bug. From a type theory standpoint, the narrowed type should be `Array[dtype_f32] & Array[T]` in this case. This can't be represented in the current type system, so it would probably need to remain unnarrowed (`Array[T]`) or be narrowed to `Array[dtype_f32]`. Neither of those would help you out in this case. If you'd like me to spend time on this, you're welcome to open a new issue.

**Comment by neutrinoceros** [comment]
thanks !
